sqlair.go

```diff
@@ -163,3 +164,80 @@ func (q *Query) One(outputArgs ...any) error {
 	iter.Decode(outputArgs...)
 	return iter.Close()
 }
+
+// All iterates over the query and decodes all rows into the provided slices.
+//
+// For example:
+//
+//	var pslice []Person
+//	var aslice []*Address
+//	err := query.All(&pslice, &aslice)
+func (q *Query) All(sliceArgs ...any) (err error) {
+	if q.err != nil {
+		return q.err
+	}
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot populate slice: %s", err)
+		}
+	}()
+
+	// Check slice inputs
+	var slicePtrVals = []reflect.Value{}
+	var sliceVals = []reflect.Value{}
+	for _, ptr := range sliceArgs {
+		ptrVal := reflect.ValueOf(ptr)
+		if ptrVal.Kind() != reflect.Pointer {
+			return fmt.Errorf("need pointer to slice, got %s", ptrVal.Kind())
+		}
+		if ptrVal.IsNil() {
+			return fmt.Errorf("need pointer to slice, got nil")
+		}
+		slicePtrVals = append(slicePtrVals, ptrVal)
+		sliceVal := ptrVal.Elem()
+		if sliceVal.Kind() != reflect.Slice {
+			return fmt.Errorf("need pointer to slice, got pointer to %s", sliceVal.Kind())
+		}
+		sliceVals = append(sliceVals, sliceVal)
+	}
+
+	iter := q.Iter()
+	for iter.Next() {
+		var outputArgs = []any{}
+		for _, sliceVal := range sliceVals {
+			elemType := sliceVal.Type().Elem()
+			var outputArg reflect.Value
+			switch elemType.Kind() {
+			case reflect.Pointer:
+				outputArg = reflect.New(elemType.Elem())
+			case reflect.Struct:
+				outputArg = reflect.New(elemType)
+			default:
+				iter.Close()
+				return fmt.Errorf("need slice of struct, got slice of %s", elemType.Kind())
+			}
+			outputArgs = append(outputArgs, outputArg.Interface())
+		}
+		if !iter.Decode(outputArgs...) {
+			break
+		}
+		for i, outputArg := range outputArgs {
+			switch sliceVals[i].Type().Elem().Kind() {
+			case reflect.Pointer:
+				sliceVals[i] = reflect.Append(sliceVals[i], reflect.ValueOf(outputArg))
+			case reflect.Struct:
+				sliceVals[i] = reflect.Append(sliceVals[i], reflect.ValueOf(outputArg).Elem())
```

Comment:
We need a `default:` here similar to the case above, but prefixed with `internal error:` as this should never happen if the prior logic is correct. This prevents awkward bugs passing through as successes and being harder to diagnose.

Reply:
Changed.